File: packages/sdk-python/devbolt/logger.py
"""Logger implementations for DevBolt."""

import json
import sys
from typing import Any, Dict, Optional

from .types import LogLevel, Logger


class NoOpLogger:
    """Logger that does nothing."""

    def debug(self, message: str, meta: Optional[Dict[str, Any]] = None) -> None:
        """Log debug message (no-op)."""
        pass

    def info(self, message: str, meta: Optional[Dict[str, Any]] = None) -> None:
        """Log info message (no-op)."""
        pass

    def warn(self, message: str, meta: Optional[Dict[str, Any]] = None) -> None:
        """Log warning message (no-op)."""
        pass

    def error(self, message: str, meta: Optional[Dict[str, Any]] = None) -> None:
        """Log error message (no-op)."""
        pass


class ConsoleLogger:
    """Console logger with configurable log level."""

    def __init__(self, level: LogLevel = LogLevel.WARN):
        """Initialize console logger.

        Args:
            level: Minimum log level to output
        """
        self.level = level

    def _format_meta(self, meta: Optional[Dict[str, Any]]) -> str:
        """Format metadata for logging."""
        if not meta:
            return ""
        return " " + json.dumps(meta, indent=2)

    def debug(self, message: str, meta: Optional[Dict[str, Any]] = None) -> None:
        """Log debug message."""
        if self.level.value <= LogLevel.DEBUG.value:
            print(f"[DevBolt Debug] {message}{self._format_meta(meta)}", file=sys.stderr)

    def info(self, message: str, meta: Optional[Dict[str, Any]] = None) -> None:
        """Log info message."""
        if self.level.value <= LogLevel.INFO.value:
            print(f"[DevBolt] {message}{self._format_meta(meta)}", file=sys.stderr)

    def warn(self, message: str, meta: Optional[Dict[str, Any]] = None) -> None:
        """Log warning message."""
        if self.level.value <= LogLevel.WARN.value:
            print(f"[DevBolt Warning] {message}{self._format_meta(meta)}", file=sys.stderr)

    def error(self, message: str, meta: Optional[Dict[str, Any]] = None) -> None:
        """Log error message."""
        if self.level.value <= LogLevel.ERROR.value:
            print(f"[DevBolt Error] {message}{self._format_meta(meta)}", file=sys.stderr)


def create_logger(level: LogLevel = LogLevel.WARN) -> Logger:
    """Create a logger instance.

    Args:
        level: Log level

    Returns:
        Logger instance
    """
    if level == LogLevel.NONE:
        return NoOpLogger()
    return ConsoleLogger(level)

File: packages/core/src/logger.ts
import { Logger, LogLevel } from "./types.js";

/**
 * No-op logger that does nothing
 */
export class NoOpLogger implements Logger {
  debug(): void {}
  info(): void {}
  warn(): void {}
  error(): void {}
}

/**
 * Console logger with configurable log level
 */
export class ConsoleLogger implements Logger {
  constructor(private readonly level: LogLevel = LogLevel.WARN) {}

  debug(message: string, meta?: Record<string, unknown>): void {
    if (this.level <= LogLevel.DEBUG) {
      console.debug(
        "[DevBolt Debug]",
        message,
        meta ? JSON.stringify(meta, null, 2) : ""
      );
    }
  }

  info(message: string, meta?: Record<string, unknown>): void {
    if (this.level <= LogLevel.INFO) {
      console.info(
        "[DevBolt]",
        message,
        meta ? JSON.stringify(meta, null, 2) : ""
      );
    }
  }

  warn(message: string, meta?: Record<string, unknown>): void {
    if (this.level <= LogLevel.WARN) {
      console.warn(
        "[DevBolt Warning]",
        message,
        meta ? JSON.stringify(meta, null, 2) : ""
      );
    }
  }

  error(message: string, meta?: Record<string, unknown>): void {
    if (this.level <= LogLevel.ERROR) {
      console.error(
        "[DevBolt Error]",
        message,
        meta ? JSON.stringify(meta, null, 2) : ""
      );
    }
  }
}

/**
 * Create a logger instance
 */
export function createLogger(level: LogLevel = LogLevel.WARN): Logger {
  if (level === LogLevel.NONE) {
    return new NoOpLogger();
  }
  return new ConsoleLogger(level);
}
